File: main/next-topological-sorting/next-topological-sorting-scratch.cpp
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <iostream>
#include <iterator>
#include <vector>

namespace {
    inline void ensure(const bool condition) noexcept
    {
        if (!condition) std::abort();
    }

    // An edge in a directed graph.
    struct Edge {
        int src;
        int dest;
    };

    inline std::istream& operator>>(std::istream& in, Edge& edge) noexcept
    {
        return in >> edge.src >> edge.dest;
    }

    // Walks through topological sorts in lexicographic order.
    class KahnBacktracker {
    public:
        // Creates a backtracker for the specified graph.
        template<typename EdgesIt, typename ChainIt>
        KahnBacktracker(int order, EdgesIt edges_first, EdgesIt edges_last,
                        ChainIt chain_first, ChainIt chain_last) noexcept;

        // Advances to the next topological sequence, if any.
        // Returns the sequence, or an empty vector if there is none.
        std::vector<int> next() noexcept;

    private:
        // Attempt to shorten the chain just enough to replace its tip with a
        // lexicographically higher vertex. If this succeeds, performs that
        // replacement. Otherwise, leaves the chain empty.
        void retreat() noexcept;

        // Lengthens the chain to produce a completed topological sort.
        void advance() noexcept;

        // Finds and removes the specified destination vertex as a root.
        void erase_root(int dest) noexcept;

        static std::size_t as_size(int count) noexcept;

        std::vector<std::vector<int>> adj_;
        std::vector<int> indegs_;
        std::vector<int> roots_;
        std::vector<int> chain_;
    };

    template<typename EdgesIt, typename ChainIt>
    KahnBacktracker::KahnBacktracker(const int order,
            const EdgesIt edges_first, const EdgesIt edges_last,
            const ChainIt chain_first, const ChainIt chain_last)
        : adj_(as_size(order) + 1u), indegs_(as_size(order) + 1u), roots_{},
          chain_(chain_first, chain_last)
    {
        const auto exists = [order](const int vertex) noexcept {
            return 0 < vertex && vertex <= order;
        };

        std::for_each(edges_first, edges_last, [&](const Edge edge) noexcept {
            ensure(exists(edge.src, edge.dest));
            adj_[src].push_back(dest);
        });

        roots_.reserve(order);

        ensure(std::all_of(cbegin(chain_), cend(chain_), exists));
        ensure(chain_.size() == order);
    }

    std::vector<int> KahnBacktracker::next() noexcept
    {
        // FIXME: implement this
    }

    void KahnBacktracker::retreat() noexcept
    {
        while (!chain_.empty()) {
            const auto src = chain_.back();
            chain_.pop_back();

            for (const auto dest : adj_[src])
                if (indegs_[dest]++ == 0) erase_root(dest);

            const auto first = begin(roots_), last = end(roots_);
            const auto src_pos = std::upper_bound(first, last, src);

            if (src_pos != last) {
                // Get the new src from roots and put the old src in its place.
                chain_.push_back(*src_pos);
                *src_pos = src;
                break;
            }

            roots_.push_back(src);
        }
    }

    void KahnBaktracker::advance() noexcept
    {
        for (const auto dest : adj_[src]) {

        }
    }

    void KahnBacktracker::erase_root(const int dest) noexcept
    {
        const auto first = begin(roots_), last = end(roots_);
        const auto dest_pos = std::lower_bound(first, last, dest);
        assert(dest_pos != last); // The root must be present.
        roots_.erase(dest_pos);
    }

    inline std::size_t KahnBacktracker::as_size(const int count) noexcept
    {
        // A graph can't have negatively many vertices, and this implementation
        // doesn't accommodate empty graphs.
        ensure(count > 0);
        return static_cast<std::size_t>(count);
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

}

File: main/next-topological-sorting/next-topological-sorting-scratch.java
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.OptionalInt;
import java.util.Scanner;
import java.util.SortedSet;
import java.util.TreeSet;

/** An edge in a directed graph. */
final class Edge {
    Edge(final int src, final int dest) {
        this.src = src;
        this.dest = dest;
    }

    final int src;
    final int dest;
}

/** Walks through topological sorts in lexicographic order. */
final class KahnBacktracker {
    /**
     * Creates a backtracker for the specified graph.
     * @param order  The number of vertices in the graph.
     * @param edges  The edges in the graph.
     * @param chain  The initial topological sort.
     */
    KahnBacktracker(final int order, final List<Edge> edges,
                    final List<Integer> chain) {
        if (order < 0) {
            throw new IllegalArgumentException(
                    "graph can't have negatively many vertices");
        }

        if (order == 0) {
            throw new IllegalArgumentException(
                    "implementation doesn't accommodate empty graph");
        }

        _adj = new ArrayList<>(order + 1);
        while (_adj.size() <= order) _adj.add(new ArrayList<>());
        for (final Edge edge : edges) _adj.get(edge.src).add(edge.dest);

        _indegs = new int[order + 1];

        _chain = new ArrayList<>(chain);
        if (_chain.size() != order) {
            throw new IllegalArgumentException(
                    "initial chain has wrong length to be a topological sort");
        }
    }

    /**
     * Advances to the next topological sequence.
     * @return The next topological sequence, if any.
     */
    Optional<List<Integer>> next() {
        // FIXME: implement this
    }

    /**
     * Attempts to shorten the chain just enough to replace its tip with a
     * lexicographically higher vertex.
     * @return The vertex that can be inserted at the tip, if any.
     */
    private OptionalInt retreat() {
        while (!_chain.isEmpty()) {
            final int src = popRootFromChain();
            final SortedSet<Integer> tail = _roots.tailSet(src + 1);

            if (!tail.isEmpty()) {
                // Put the new src in roots and return the old src.
                final int res = tail.first();
                tail.remove(res);
                _roots.add(src);
                return OptionalInt.of(res);
            }

            _roots.add(src);
        }

        return OptionalInt.empty();
    }

    /**
     * Pops a vertex from the chain for subsequent insertion as a root, and
     * increments its forward neighbors' indegrees.
     * @return  The popped vertex, which was previously the tip of the chain.
     */
    private int popRootFromChain() {
        final int src = _chain.remove(_chain.size() - 1);

        for (final int dest : _adj.get(src))
            if (_indegs[dest]++ == 0) _roots.remove(dest);

        return src;
    }

    /**
     * Pushes a presumably just-extracted root to the chain.
     * Decrements its forward neighbors' indegrees.
     * @param src  The extracted root, which will be the new tip of the chain.
     */
    private void pushRootToChain(final int src) {
        for (final int dest : _adj.get(src))
            if (--_indegs[dest] == 0) _roots.add(dest);

        _chain.add(src);
    }

    private final SortedSet<Integer> _roots = new TreeSet<>();
    private final List<List<Integer>> _adj;
    private final int[] _indegs;
    private final List<Integer> _chain;
}
